warehouse/query-core/src/main/java/datawave/query/jexl/visitors/whindex/WhindexVisitor.java

```diff
@@ -0,0 +1,974 @@
+import datawave.query.jexl.visitors.QueryPropertyMarkerVisitor;
+import datawave.query.jexl.visitors.RebuildingVisitor;
+import datawave.query.jexl.visitors.TreeFlatteningRebuildingVisitor;
+import datawave.query.util.MetadataHelper;
+import datawave.util.time.DateHelper;
+import datawave.webservice.common.logging.ThreadConfigurableLogger;
+import org.apache.commons.jexl2.parser.ASTAndNode;
+import org.apache.commons.jexl2.parser.ASTDelayedPredicate;
+import org.apache.commons.jexl2.parser.ASTEQNode;
+import org.apache.commons.jexl2.parser.ASTERNode;
+import org.apache.commons.jexl2.parser.ASTFunctionNode;
+import org.apache.commons.jexl2.parser.ASTGENode;
+import org.apache.commons.jexl2.parser.ASTGTNode;
+import org.apache.commons.jexl2.parser.ASTLENode;
+import org.apache.commons.jexl2.parser.ASTLTNode;
+import org.apache.commons.jexl2.parser.ASTNENode;
+import org.apache.commons.jexl2.parser.ASTNRNode;
+import org.apache.commons.jexl2.parser.ASTNotNode;
+import org.apache.commons.jexl2.parser.ASTOrNode;
+import org.apache.commons.jexl2.parser.ASTReference;
+import org.apache.commons.jexl2.parser.ASTReferenceExpression;
+import org.apache.commons.jexl2.parser.JexlNode;
+import org.apache.commons.jexl2.parser.JexlNodes;
+import org.apache.log4j.Logger;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * The 'WhindexVisitor' is used to replace wide-scoped geowave fields with value-specific, narrow-scoped geowave fields where appropriate.
+ * <p>
+ * For example, assume you have a wide-scope geowave field (e.g. GEOWAVE_FIELD), a narrow-scope geowave field(s) (e.g. MARS_GEOWAVE_FIELD or
+ * EARTH_GEOWAVE_FIELD), and a separate field whose value is specific to each narrow-scoped field (e.g. PLANET == 'MARS' or PLANET == 'EARTH').
+ * <p>
+ * If a query comes in with the following form: geowave:intersects(GEOWAVE_FIELD, '...some wkt...) &amp;&amp; PLANET == 'MARS'
+ * <p>
+ * The WhindexVisitor would turn that into this: geowave:intersects(MARS_GEOWAVE_FIELD, '...some wkt')
+ * <p>
+ * The update function uses the value-specific geowave field, and drops the anded term (since all of the data indexed under the value-specific geowave field
+ * already satisfies that term).
+ *
+ */
+public class WhindexVisitor extends RebuildingVisitor {
+    private static final Logger log = ThreadConfigurableLogger.getLogger(WhindexVisitor.class);
+    
+    private final Set<String> mappingFields;
+    private final Map<String,Map<String,String>> valueSpecificFieldMappings;
+    private final MetadataHelper metadataHelper;
+    
+    private final HashMap<JexlNode,WhindexTerm> jexlNodeToWhindexMap = new HashMap<>();
+    
+    private static class ExpandData {
```

Comment:
can you describe how these are used/maintained in the recursion?

Reply:
done